### ISSUE TITLE
Text parser mistakenly supports missing comma between label terms in the selector

### ISSUE THREAD

**Issue by ywwg** [prometheus/prometheus#13273]
scrape_test.go contains a few tests that contain spaces between label terms: 
https://github.com/prometheus/prometheus/blob/main/scrape/scrape_test.go#L1330
and
https://github.com/prometheus/prometheus/blob/main/scrape/scrape_test.go#L1405

This is not allowed in OpenMetrics parsing and does not appear to be intended behavior given the definition of the format as 

```
metric_name [
  "{" label_name "=" `"` label_value `"` { "," label_name "=" `"` label_value `"` } [ "," ] "}"
] value [ timestamp ]
```

This can be fixed but should probably wait until Prom 3.0

**Comment by beorn7** [comment]
Historically, the Prometheus text format was created to allow very simple expositions, even makeshift ones (curl a metric or something). With that in mind, the format was already formulated quite tolerantly (whitespace allow almost everywhere etc.), but the parsers was made even more tolerant, arguably following the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle). The danger of that is that people might test their exposition against Prometheus, and then will wake up to a bad surprise if another scraper tries to parse their exposition. The historical viewpoint was that "proper" exposition would be done with the protobuf format anyway. That changed in between, but now (with more efficient protobuf code in Go and native histograms) we might see more protobuf usage again.

It was always my opinion that the strictness in the OpenMetrics text format is a very problematic design flaw (for reasons explained elsewhere). However, my worries are more about things like whitespaces, that were part of the Prometheus text format, not the additional tolerance that the parser granted. (Note, however, that even the OpenMetrics parser in Prometheus tolerates a few things that are not specified in OpenMetrics, which has already caused confusion. I generally prefer to have tolerance in the format itself.)

Long story short: Ideally, the future is OpenMetrics (hopefully in an improved v2.0). Support of the old Prometheus text format will probably have to stay to not break the many legacy endpoints. But for the same reason, I would not change the parser's behavior for it, unless we have to.

**Comment by bboreham** [comment]
So we close this as "works as designed" ?

**Comment by beorn7** [comment]
"As designed" might be a bit strong, but I would indeed suggest to not change the current behavior, unless it gets in the way of something else (like the problems we ran into with the overly tolerant matcher parser in Alertmanager).

**Comment by bboreham** [comment]
It's not about the spaces, it's about having just a space and no comma.
Since the code has moved, the links at the top don't show the problem. Here are links with fixed commit:
https://github.com/prometheus/prometheus/blob/9c4782f1cc34f85f07daaaed14af2e7ad1ff56a0/scrape/scrape_test.go#L1330
https://github.com/prometheus/prometheus/blob/9c4782f1cc34f85f07daaaed14af2e7ad1ff56a0/scrape/scrape_test.go#L1405

**Comment by sagar-cs03** [comment]
@carrieedwards are we going to deprecate the old prometheus metric format in favour of OpenMetric format in prometheus 3.0 ? Can I work on it ?

**Comment by beorn7** [comment]
I think we need to make a call here. I expressed my opinion above with the rational behind it. If we follow it, we can close this issue (as we will keep the current behavior for the legacy text format parser). Working to improve OpenMetrics and make it the actual de-facto standard is a different thing out of scope of this issue.

**Comment by bboreham** [comment]
I agree with Beorn's reasoning, so we can close this issue.
I don't know what the OpenMetrics improvement would be; someone who does can type that in.
